docs/defaulting.md

```diff
@@ -0,0 +1,249 @@
+isolated libvirt (running libvirt and the small REST service in a Pod). The
+service is a standalone component in KubeVirt and can be integrated in the
+following flow:
+
+ * When doing a POST to
+   `/apis/kubevirt.io/v1alpha1/namespaces/mynamespace/vms`, `virt-api` can do
+   a roundtrip to that service and prefill the defaults. As the response of
+   the POST to `virt-api`, you will get the completely prefilled VM
+   specification. This updated spec contains all the necessary bits, to allow
+   consistent hotplugging and keeping the VM definition consistent between
+   redefines.
+
+Additional flows, like a dedicated `dry-run` endpoint, to allow management
+applications, reusing existing admission controller logic for prefilling VMs,
+withouth having to actually run a VM are thinkable. That would allow complete
+decoupling of managment and runtime.
+
+Defaulting rules:
+
+ * The Kubevirt VM spec can contain abstract definitions, like a reference to a
+   volume claim, or a reference to a host network. In such cases most of the
+   time, the device `source` section needs to be replaced by a dummy `source`
+   inside the defaulter. Then the fully populated `target` section needs to be
+   mapped back (e.g. a PVC does not have to exist, when posting a VM. It can be
+   mocked out, since it only provides the backend configuration).
+ * Fields which don't influence how the Guest OS sees, shall not be defaulted.
+   Such fields need to be filtered out.
+ * There exist mandatory Domain XML fields, which we can't allow to be set on
+   the VM specification. In such cases, the defaulter needs to set dummy values
+   before asking libvirt for the defaults. Since these fields should not exist
+   on the VM specification at all (see
+   [docs/vm-configuration.md](docs/vm-configuration.md)), not mapping such
+   dummy data back should be easy.
+
+## Implementation Details
+
+The implementation consists of two parts. First the libvirt
+[admission-controller](https://kubernetes.io/docs/admin/admission-controllers/#what-are-they)
+in `virt-api`. This admission controller, then reaches out to the second part,
+a special libvirt instance in a container, wrapped by a REST API, which acts as
+libvirt defaulting service. Since the defaults from this service would normally
+be filled in on the host, it is important that this controller is the last of
+all invoked admission controllers. For instance, if another admission
+controller would exist, which allows injecting Disks, like a
+[PodPreset](https://kubernetes.io/docs/tasks/inject-data-application/podpreset/)
+allows injecting volumes, it has to be invoked before this libvirt defaulting
+service. Otherwise relevant specification details, like PCI addresses might
+still be added at the host level. the final flow looks like this:
+
+```
+ client       virt-api/apiserver admission-controller 1 libvirt admission controller  libvirt defaulting service
```

Comment:
it's not just fancy stuff, it's elementary stuff, e.g. virtio-scsi controller in your example above. Libvirt defaults are as bad as QEMU defaults - sadly it's not what any sane product would want to use.

Anyway, I think it sounds ok, I just think we should have started with the admission controllers before the lowest-level libvirt defaulting.